src/build_vehicle_gui.cpp

```diff
@@ -1427,58 +1426,58 @@ struct BuildVehicleWindow : Window {
 
 		/* and finally sort wagons */
 		EngList_SortPartial(list, _engine_sort_functions[0][this->sort_criteria], num_engines, list.size() - num_engines);
+
+		return sel_id;
 	}
 
 	/* Figure out what road vehicle EngineIDs to put in the list */
 	void GenerateBuildRoadVehList()
 	{
-		EngineID sel_id = INVALID_ENGINE;
-
 		this->eng_list.clear();
+		this->num_hidden_engines = 0;
 
 		for (const Engine *e : Engine::IterateType(VEH_ROAD)) {
-			if (!this->show_hidden_engines && e->IsVariantHidden(_local_company)) continue;
 			EngineID eid = e->index;
 			if (!IsEngineBuildable(eid, VEH_ROAD, _local_company)) continue;
 			if (this->filter.roadtype != INVALID_ROADTYPE && !HasPowerOnRoad(e->u.road.roadtype, this->filter.roadtype)) continue;
+			if (e->IsVariantHidden(_local_company)) {
+				this->num_hidden_engines++;
+				if (!this->show_hidden_engines) continue;
+			}
 
 			/* Filter by name or NewGRF extra text */
 			if (!FilterByText(e)) continue;
 
 			this->eng_list.emplace_back(eid, e->info.variant_id, e->display_flags, 0);
-
-			if (eid == this->sel_engine) sel_id = eid;
 		}
-		this->SelectEngine(sel_id);
```

Comment:
This function, and its counterparts below for ships and aircrafts, don't follow the `GenerateBuildTrainList()` example above, which removes this call but adds a return. Why? Ideally the doxygen would better explain what these functions actually do, because (no fault of yours) they are somewhat hard to read and understand.

Reply:
Well, the answer is in GenerateBuildList, trains are a special case that do their own cargo filtering while other vehicle types use FilterEngineList that also adds this return.